app/src/main/java/com/github/se/icebreakrr/model/tags/TagsRepository.kt

```diff
@@ -0,0 +1,250 @@
+      db.collection(collectionPath)
+          .document(categoryName)
+          .get()
+          .addOnSuccessListener { documentSnapshot: DocumentSnapshot ->
+            var tagsCategory = TagsCategory()
+            firestoreToTags(
+                documentSnapshot,
+                { tagCategoryCallback -> tagsCategory = tagCategoryCallback },
+                { e: Exception ->
+                  Log.e(
+                      "TagsRepository",
+                      "[addCategory] error while converting firebase to Tags : $e")
+                  onFailure(e)
+                })
+            val subtags = tagsCategory.subtags.toMutableList()
+            subtags.addAll(subcategories.filter { it !in subtags })
+            db.collection(collectionPath)
+                .document(categoryName)
+                .update("subtags", subtags.toList())
+                .addOnSuccessListener {}
+                .addOnFailureListener { exception: Exception ->
+                  Log.e("TagsRepository", "[addCategory] Could not update the document: $exception")
+                  onFailure(exception)
+                }
+          }
+          .addOnFailureListener { e: Exception ->
+            Log.e("TagsRepository", "[addCategory] fail to get the document :  $e")
+            onFailure(e)
+          }
+    } else {
+      db.collection(collectionPath)
+          .document(categoryName)
+          .set(TagsCategory(categoryName, color, subcategories))
+          .addOnSuccessListener {}
+          .addOnFailureListener { e: Exception ->
+            Log.e("TagsRepository", "[addCategory] could not set the new category :  $e")
+            onFailure(e)
+          }
+    }
+  }
+
+  /**
+   * use this function to delete a tag in a certain category. If the tag isn't in the category, or
+   * the category doesn't exist, the method does nothing.
+   *
+   * @param onFailure : happens if the tags can't be fetch from the database or if the update of the
+   *   database fails
+   * @param name : name of the tag to delete
+   * @param category : category in which the tag you want to suppress is
+   */
+  fun deleteTag(onFailure: (Exception) -> Unit, name: String, category: CategoryString) {
+    var tagList = emptyList<TagsCategory>()
+    getAllTags(
+        { tagList = it }, { Log.e("TagsRepository", "[deleteTag] error while getting tags : $it") })
+    val tagCategory =
+        try {
+          tagList.filter { it.name == category.name }[0]
+        } catch (e: Exception) {
+          null
+        }
+    if (tagCategory != null && tagCategory.subtags.contains(name)) {
+      db.collection(collectionPath)
+          .document(category.name)
+          .update("subtags", tagCategory.subtags.filter { it != name })
+          .addOnSuccessListener {}
+          .addOnFailureListener { e: Exception ->
+            Log.e("TagsRepository", "[deleteTag] Failed to update the document : $e")
+            onFailure(e)
+          }
+    }
```

Comment:
Problem (ft. copilot): there is a problem of mixing synchronous and asynchronous code here. `getAllTags()` is asynchronous as it does network operation. this means `tagList `may not be populated yet if you are using it just after the `getAllTags()` block.

Here is a solution proposed by copilot (to show you how asynchronous code should be used)
```

fun deleteTag(onFailure: (Exception) -> Unit, name: String, category: CategoryString) {
  getAllTags(
      { tagList ->
        val tagCategory = tagList.find { it.name == category.name }
        if (tagCategory != null && tagCategory.subtags.contains(name)) {
          db.collection(collectionPath)
              .document(category.name)
              .update("subtags", tagCategory.subtags.filter { it != name })
              .addOnSuccessListener {}
              .addOnFailureListener { e: Exception ->
                Log.e("TagsRepository", "[deleteTag] Failed to update the document : $e")
                onFailure(e)
              }
        } else {
          Log.d(
              "TagsRepository",
              "[deleteTag] the tag wasn't in the database or the category doesn't exist : $name")
        }
      },
      { e ->
        Log.e("TagsRepository", "[deleteTag] error while getting tags : $e")
        onFailure(e)
      }
  )
}
```